plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,317 @@
+      defaultText = literalExpression "config.plugins.treesitter.package.passthru.allGrammars";
+      description = "Grammar packages to install";
+    };
+
+    languageRegister = mkOption {
+      type = with types; attrsOf (coercedTo str toList (listOf str));
+      default = { };
+      example = {
+        cpp = "onelab";
+        python = [
+          "foo"
+          "bar"
+        ];
       };
+      description = ''
+        This is a wrapping of the `vim.treesitter.language.register` function.
+        Register specific parsers to one or several filetypes.
+        The keys are the parser names and the values are either one or several filetypes.
+      '';
+    };
+
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = "Whether to install grammars defined in `grammarPackages`.";
+    };
+
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = ''Whether to enable Nixvim injections, such as recognising extraConfigLua as lua.'';
+    };
 
-      nixvimInjections = mkEnableOption "nixvim specific injections, like lua highlighting in extraConfigLua";
+    nodejsPackage = helpers.mkPackageOption {
+      name = "nodejs";
+      default = pkgs.nodejs;
+      defaultText = literalExpression "pkgs.nodejs";
+      example = literalExpression "pkgs.nodejs_22";
+      description = ''
+        Which package (if any) to be added as the nodejs package.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of NodeJS, set this option to `null`.
+      '';
+    };
+
+    treesitterPackage = helpers.mkPackageOption {
+      name = "tree-sitter";
+      default = pkgs.tree-sitter;
+      defaultText = literalExpression "pkgs.tree-sitter";
+      description = ''
+        Which package (if any) to be added as the treesitter binary.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of treesitter, set this option to `null`.
+      '';
     };
   };
 
-  config =
-    let
-      tsOptions = {
-        highlight = {
-          inherit (cfg) enable;
-          disable = if (cfg.disabledLanguages != [ ]) then cfg.disabledLanguages else null;
+  # NOTE: We call setup manually below.
+  callSetup = false;
+  # NOTE: We install cfg.package manually so we can install grammars using it.
+  installPackage = false;
 
-          custom_captures = if (cfg.customCaptures != { }) then cfg.customCaptures else null;
-        };
+  extraConfig = cfg: {
+    assertions =
+      map
+        (name: {
+          assertion = !cfg.nixGrammars -> cfg.${name} != null;
+          message = ''
+            Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.${name}` is `null`.
+            This package is required to build the grammars at runtime.
+          '';
```

Comment:
Does this need to be an assertion or is a warning enough?

Is it foreseeable that a user would want to set these packages to null and also disable `nixGrammars`, but not compile any grammars at runtime?

If so, I think we want a warning, otherwise the assertion is correct.

Reply:
Hmm.. maybe along the same lines of thinking that they managed the packages outside of nixvim. Like we talked about with the main package binary being null.

Reply:
Unless you (or anyone) disagrees, I'm fine with the assertion. We can always re-visit if someone complains.

If you did want to do a warning, the schema is slightly different; it is a simple list of warning strings.\
Therefore, we need to use `optional`, `filter` or `mkIf` to decide whether or not each warning should be set.